src/main/java/bio/terra/workspace/db/DataReferenceDao.java

```diff
@@ -0,0 +1,152 @@
+          .resourceDescription(
+              rs.getString("resource_id") == null
+                  ? null
+                  : resourceDescriptionMapper.mapRow(rs, rowNum))
+          .credentialId(rs.getString("credential_id"))
+          .cloningInstructions(
+              CloningInstructionsEnum.fromValue(rs.getString("cloning_instructions")))
+          .referenceType(ReferenceTypeEnum.fromValue(rs.getString("reference_type")))
+          .reference(rs.getString("reference"));
+    }
+  }
+
+  // Returns a SQL condition as a string that selects controlled resources, uncontrolled resources,
+  // or both, depending on the value of filterControlled.
+  public static String filterControlledWhereClause(
+      FilterControlledEnum filterControlled, String referenceTableAlias) {
+    // filterControlled is one of "controlled", "uncontrolled", or "all" (enum in OpenAPI yaml).
+    // This corresponds to the workspace_data_reference.resource_id column being not null, null, or
+    // either, respectively.
+    switch (filterControlled) {
+      case CONTROLLED:
+        return "(" + referenceTableAlias + ".resource_id IS NOT NULL)";
+      case UNCONTROLLED:
+        return "(" + referenceTableAlias + ".resource_id IS NULL)";
+      case ALL:
+        return StringUtils.EMPTY;
+      default:
+        // This case should not happen, as OpenAPI should enforce that this value is an enum.
+        // Most likely cause is that a new enum value is added to the definition but not
+        // added here.
+        throw new ValidationException(
+            "Invalid value for filterControlled: "
+                + filterControlled
+                + " in filterControlledWhereClause. Has a new value been added to FilterControlledEnum?");
+    }
+  }
+
+  // Returns a SQL condition as a string accepts both uncontrolled data references and visible
+  // controlled references. Uncontrolled references are not tracked as resources, and their
+  // existence is always visible to all workspace readers.
+  public static String uncontrolledOrVisibleResourcesClause(
+      String owner, String resourceTableAlias, String referenceTableAlias) {
+    return "(("
+        + referenceTableAlias
+        + ".resource_id IS NULL) OR "
+        + visibleResourcesClause(owner, resourceTableAlias)
+        + ")";
+  }
+
+  // Returns a SQL condition as a string that filters out invisible controlled references.
+  // References are considered 'invisible' if the is_visible column of the corresponding resource
+  // is false AND the resource owner is not the user issuing the query.
+  // Uncontrolled references are not tracked as resources, and their existence is always visible
+  // to all workspace readers.
+  public static String visibleResourcesClause(String owner, String resourceTableAlias) {
+    return "("
+        + resourceTableAlias
+        + ".is_visible = true OR "
+        + resourceTableAlias
+        + ".owner = \'"
+        + owner
+        + "\')";
+  }
```

Comment:
this should use `:owner` syntax and SQL parameter replacement, else it is vulnerable to injection

Reply:
Good catch!